README.rst
===============================
Sex Talk
===============================

.. image:: https://badge.fury.io/py/sextalk.png
    :target: http://badge.fury.io/py/sextalk
    
.. image:: https://travis-ci.org/judowill/sextalk.png?branch=master
        :target: https://travis-ci.org/judowill/sextalk

.. image:: https://pypip.in/d/sextalk/badge.png
        :target: https://crate.io/packages/sextalk?version=latest


The analysis code, figures, and slides for my QS Sex Talk.

* Free software: BSD license
* Documentation: http://sextalk.rtfd.org.

Features
--------

* TODO

HISTORY.rst
.. :changelog:

History
-------

0.1.0 (0.1dev)
++++++++++++++++++

* First release on PyPI.

notebooks/StressAnalysis.py
# -*- coding: utf-8 -*-
# <nbformat>3.0</nbformat>

# <headingcell level=1>

# Gmail Stress Measure

# <markdowncell>

# Since I use Gmail as my main work communication, I can reasonably assume that the number of work-related emails is a proxy for the amount of work-related stress I'm under at any given time. I used the `imaplib` tool in the python standard library to extract emails from my Gmail account. I then used `pandas` to aggregate the data.

# <codecell>

import imaplib
imap_server = imaplib.IMAP4_SSL("imap.gmail.com",993)
imap_server.login('judowill', open('SUPERSECRET.txt').read())
imap_server.select('[Gmail]/All Mail')

# <markdowncell>

# These are the emails of people that I work with. I assume that most emails to or from these people are work related. So to avoid putting them on the internet, they're saved in a file not present in Github. But if you make a TSV file of Email-Name pairs then the code will work for you as well.

# <codecell>

import csv
work_related_emails = list(csv.reader(open('SECRETEMAILS.tsv'), delimiter = '\t'))

# <codecell>

import dateutil.parser
import email


def get_emails_person(address):
    search = '(OR (TO "%(name)s") (FROM "%(name)s"))'
    status, response = imap_server.search(None, search % {'name':address})
    ids = response[0].split()
    for _id in ids:
        yield _id
        
def get_date_from_ids(ids):
    
    resp, data = imap_server.FETCH(','.join(ids), 'RFC822.HEADER')
    id_iter = iter(ids)
    for d in data:
        try:
            msg = email.message_from_string(d[1])
            try:
                tdate = dateutil.parser.parse(msg['DATE'])
            except TypeError:
                continue
            yield id_iter.next(), tdate
        except (IndexError, ValueError):
            pass
        
from itertools import islice

def take(n, iterable):
    "Return first n items of the iterable as a list"
    return list(islice(iterable, n))

known_dates = {}

# <codecell>

email_res = []

for em, name in work_related_emails:
    print name
    needed_ids = []
    already_have = 0
    for _id in get_emails_person(em):
        if _id in known_dates:
            already_have += 1
            email_res.append({
                          'DateTime': known_dates[_id],
                          'Address':em,
                          'Name': name,
                          'EmailID':_id
                          })
        else:
            needed_ids.append(_id)
    print 'already had', already_have
    needed_iter = iter(needed_ids)
    block = take(50, needed_iter)
    count = 0
    while len(block)>0:
        if count % 10 == 0:
            print 'block', count, 'of', len(needed_ids)/50
        for _id, tdate in get_date_from_ids(block):
            known_dates[_id] = tdate
            email_res.append({
                          'DateTime': known_dates[_id],
                          'Address':em,
                          'Name': name,
                          'EmailID':_id
                          })
                
        count += 1
        block = take(50, needed_iter)

# <codecell>

import pandas as pd

email_df = pd.DataFrame(email_res)
email_df['nDateTime'] = pd.to_datetime(email_df['DateTime'], utc=True)

grouped_emails = email_df.groupby(['Name', 'nDateTime'])['EmailID'].count()
all_emails = email_df.groupby('nDateTime')['EmailID'].count().resample('D', how = 'sum').fillna(0)

# <codecell>

fig, axs = plt.subplots(2,1, sharex=True, sharey=True, figsize=(10,10))

groups = [('Aydin', 'Ertel', 'Yichuan', 'Perry', 'Gormley', 'Noor', 'Ceylan', 'Mahdi'),
          ('BrianW', 'Mike', 'Vanessa', 'Niz', 'Benj')]

for names, ax in zip(groups, axs.flatten()):
    for name in names:
        named_emails = grouped_emails.ix[name].resample('D', how='sum').resample('7D', how = 'sum')
        t, large_named_emails = all_emails.align(named_emails, join = 'left')
        large_named_emails.fillna(0).plot(ax = ax, label = name, lw=4, alpha=0.5)
    ax.legend(loc = 'best')
    ax.set_ylabel('#Emails/Week')
    
fig.savefig('figures/EmailsPerWeek.png', dpi = 600)

# <markdowncell>

# I believe that my stress isn't just the number of emails I get per day, but how that number of emails compares to previosuly stressful days. This shows an expanding-max showing the concept.

# <codecell>

fig, ax = plt.subplots(1,1, figsize = (10, 5))

max_emails = pd.expanding_max(all_emails)
all_emails.plot(ax=ax, color = 'b', label = 'All')
max_emails.plot(ax=ax, color = 'b', lw = 10, alpha = 0.1, label = 'Expanding Max')
ax.set_ylabel('Emails/Day')
ax.legend(loc = 'best')
fig.savefig('figures/ExpandingMax.png', dpi = 600)

# <codecell>

fig, axs = plt.subplots(3,1, sharex=True, figsize=(10,15))

for names, ax in zip(groups, axs.flatten()):
    for name in names:
        named_emails = grouped_emails.ix[name].resample('D', how='sum').resample('7D', how = 'sum')
        t, large_named_emails = all_emails.align(named_emails, join = 'left')
        large_named_emails.fillna(0).plot(ax = ax, label = name)
    #ax.legend(loc = 'best')
    ax.set_ylabel('#Emails/Week')

stress = (all_emails/max_emails)
stress.resample('7D', how = 'mean').plot(ax = axs[-1])
axs[-1].set_ylabel('Stress -- Weekly Avg')
fig.savefig('figures/EmailsPlusStress.png', dpi = 600)

# <codecell>

pd.DataFrame({'Stress': stress}).to_csv('results/stress.tsv', sep = '\t')

# <codecell>




notebooks/PreferenceAnalysis.py
# -*- coding: utf-8 -*-
# <nbformat>3.0</nbformat>

# <codecell>

from __future__ import division
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import poisson
from functools import partial
from datetime import datetime, timedelta

# <codecell>

sex_data = pd.read_csv('projectdata/willdata.tsv', sep = '\t', parse_dates = [0], index_col=0)
sex_data.head()

# <codecell>

stress_data = pd.read_csv('results/stress.tsv', sep = '\t', parse_dates = [0])

# <codecell>

final_data = pd.merge(sex_data, stress_data.groupby('nDateTime')[['Stress']].first(),
                      left_index=True, 
                      right_index = True,
                      how = 'left')
final_data.head()

         

# <codecell>

fig, ax = plt.subplots(3,1, figsize = (15, 10))

pd.rolling_mean(final_data.any(axis=1).resample('W', how='sum'), 4, min_periods = 1).plot(ax = ax[0])
ax[0].set_ylabel('#/week')

week_counts = final_data[['missionary', 'cowgirl', 'doggy', 'kinky']].resample('W', how = 'sum')
pd.rolling_mean(week_counts, 2, min_periods=1).plot(ax = ax[1])
ax[1].set_ylabel('#/week')
ax[1].set_ylim(0, 10);

week_ratings = final_data[['HerRanking', 'MyRanking']].resample('W', how='median')
pd.rolling_median(week_ratings, 4, min_periods = 3).plot(ax = ax[2])
ax[2].set_ylabel('Mean Satisfaction')
ax[2].set_ylim(4, 10);
fig.savefig('figures/frequency.png', dpi = 400)

# <codecell>

from statsmodels.tools.tools import maybe_unwrap_results
from statsmodels.graphics import utils
import statsmodels.formula.api as smf

def ccpr_data(results, exog_idx):
    
    exog_name, exog_idx = utils.maybe_name_or_idx(exog_idx, results.model)
    results = maybe_unwrap_results(results)
    
    x1 = results.model.exog[:, exog_idx]
    x1beta = x1*results.params[exog_idx]
    
    return x1, x1beta + results.resid
    

# <codecell>

from scipy.stats import linregress

eqns = [('b', 'MyRanking ~ PreOrgs + missionary + doggy + cowgirl + reversecowgirl + kinky + Stress'),
        ('r', 'HerRanking ~ PreOrgs + missionary + doggy + cowgirl + reversecowgirl + kinky + Stress')]
plot_cols = [('PreOrgs', 'Cont'),  
             ('Stress', 'Cont'),
             ('missionary', 'bool'), 
             ('doggy', 'bool'), 
             ('cowgirl', 'bool'), 
             ('reversecowgirl', 'bool'), 
             ('kinky', 'bool')]
groups = ['J1', 'J2', 'M']

fig, axs = plt.subplots(len(plot_cols), 3, figsize = (20, 20))

for color, endog in eqns:
    
    for axcol, g in enumerate(groups):
        tdata = final_data.query("Name == '%s'" % g)
        res = smf.ols(endog, data = tdata).fit()
        for ax, (val, typ) in zip(axs[:,axcol].flatten(), sorted(plot_cols)):
            #try:
            x, y = ccpr_data(res, val)
            #except ValueError:
            #    print 'bad error'
            #    continue
                
            if ax.is_first_row():
                ax.set_title(g)
            
            if typ == 'bool':
                tmp = pd.DataFrame({'X': x, 'y':y})
                try:
                    tmp.query('X == 1')['y'].plot(kind = 'kde', 
                                                  ax = ax, 
                                                  linewidth = 8, 
                                                  alpha = 0.4,
                                                  color = color, 
                                                  grid=False)
                except:
                    pass
                ax.set_xlim(-10, 10)
                if ax.is_last_row():
                    ax.set_xlabel('Rating')
                ax.set_ylabel('')
                ax.set_title(val)
                ax.vlines(0, *ax.get_ylim())
                ax.set_ylim(0, ax.get_ylim()[1])
            else:
                ax.scatter(x, y, alpha = 0.1, color = color, s=30)
                tup = linregress(x, y)
                m, b = tup[0], tup[1]
                xp = np.linspace(x.min(), x.max())
                ax.plot(xp, m*xp+b, color = color, linewidth = 10, alpha = 0.5)
                ax.set_ylim(-5, 5)
                ax.set_xlabel(val)
                ax.set_ylabel('Rating')
fig.tight_layout()
fig.savefig('figures/position_preference.png', dpi = 600)

# <codecell>

eqns = [('b', 'Me', 'MyRanking ~ PreOrgs + missionary + doggy + cowgirl + reversecowgirl + kinky + Stress'),
        ('r', 'Her', 'HerRanking ~ PreOrgs + missionary + doggy + cowgirl + reversecowgirl + kinky + Stress')]
#plot_cols = set(['PreOrg', 'missionary[T.True]', 'doggy[T.True]', 'cowgirl[T.True]', 'reversecowgirl[T.True]', 'kinky[T.True]'])
groups = ['J1', 'J2', 'M']

tres = []
for _, per, eqn in eqns:
    for g in groups:
        tdata = final_data.query("Name == '%s'" % g)
        res = smf.ols(eqn, data = tdata).fit()
        tmp = res.params.copy()
        tmp['Name'] = g
        tmp['Person'] = per
        tres.append(tmp.copy())

res_df = pd.DataFrame(tres)

# <codecell>

import matplotlib as mpl

tcols = sorted([c for c, _ in plot_cols])

cmap = mpl.cm.BrBG
norm = mpl.colors.Normalize(vmin=-5, vmax=5)

fig, axs = plt.subplots(3,1, figsize = (10, 10))
axs[1].imshow(res_df.query('Person == "Her"')[tcols].values,
           interpolation='nearest', norm=norm, aspect='auto', cmap = cmap)
axs[1].set_yticks([0, 1, 2])
axs[1].set_yticklabels(groups)
axs[1].set_title('Her Effect')
axs[1].set_xticks([])

axs[2].imshow(res_df.query('Person == "Me"')[tcols].values,
           interpolation='nearest', norm=norm, aspect='auto', cmap = cmap)
axs[2].set_yticks([0, 1, 2])
axs[2].set_yticklabels(groups)
axs[2].set_xticks(range(len(tcols)))
axs[2].set_xticklabels(tcols, rotation = 90)
axs[2].set_title('My Effect')

cb3 = mpl.colorbar.ColorbarBase(axs[0], cmap=cmap,
                                     norm=norm,
                                     extend='both',
                                     # Make the length of each extension
                                     # the same as the length of the
                                     # interior colors:
                                     extendfrac='auto',
                                     spacing='uniform',
                                     orientation='horizontal')
fig.tight_layout()
fig.savefig('figures/effect_heatmap.png', dpi = 400)

# <codecell>




.config/ipython/profile_default/ipython_notebook_config.py
# Configuration file for ipython-notebook.

c = get_config()

#------------------------------------------------------------------------------
# NotebookApp configuration
#------------------------------------------------------------------------------

# NotebookApp will inherit config from: BaseIPythonApplication, Application

# The IPython profile to use.
# c.NotebookApp.profile = 'default'
c.IPKernelApp.pylab = 'inline'

# The url for MathJax.js.
# c.NotebookApp.mathjax_url = ''

# The IP address the notebook server will listen on.
c.NotebookApp.ip = '*'

# The base URL for the notebook server
c.NotebookApp.base_project_url = '/sextalk/'

# Create a massive crash report when IPython encounters what may be an internal
# error.  The default is to append a short message to the usual traceback
# c.NotebookApp.verbose_crash = False

# The number of additional ports to try if the specified port is not available.
# c.NotebookApp.port_retries = 50

# Whether to install the default config files into the profile dir. If a new
# profile is being created, and IPython contains config files for that profile,
# then they will be staged into the new directory.  Otherwise, default config
# files will be automatically generated.
# c.NotebookApp.copy_config_files = False

# The base URL for the kernel server
c.NotebookApp.base_kernel_url = '/sextalk/'

# The port the notebook server will listen on.
c.NotebookApp.port = 8888

# Whether to overwrite existing config files when copying
# c.NotebookApp.overwrite = False

# Whether to prevent editing/execution of notebooks.
# c.NotebookApp.read_only = False

# Whether to enable MathJax for typesetting math/TeX
# 
# MathJax is the javascript library IPython uses to render math/LaTeX. It is
# very large, so you may want to disable it if you have a slow internet
# connection, or for offline use of the notebook.
# 
# When disabled, equations etc. will appear as their untransformed TeX source.
c.NotebookApp.enable_mathjax = True

# Whether to open in a browser after starting. The specific browser used is
# platform dependent and determined by the python standard library `webbrowser`
# module, unless it is overridden using the --browser (NotebookApp.browser)
# configuration option.
# c.NotebookApp.open_browser = True

# The full path to an SSL/TLS certificate file.
# c.NotebookApp.certfile = ''

# The hostname for the websocket server.
# c.NotebookApp.websocket_host = ''

# The name of the IPython directory. This directory is used for logging
# configuration (through profiles), history storage, etc. The default is usually
# $HOME/.ipython. This options can also be specified through the environment
# variable IPYTHONDIR.
# c.NotebookApp.ipython_dir = '/home/will/.config/ipython'

# Set the log level by value or name.
# c.NotebookApp.log_level = 20

# Hashed password to use for web authentication.
# 
# To generate, type in a python/IPython shell:
# 
#   from IPython.lib import passwd; passwd()
# 
# The string should be of the form type:salt:hashed-password.
# c.NotebookApp.password = ''

# The Logging format template
# c.NotebookApp.log_format = '[%(name)s] %(message)s'

# The full path to a private key file for usage with SSL/TLS.
# c.NotebookApp.keyfile = ''

# Supply overrides for the tornado.web.Application that the IPython notebook
# uses.
# c.NotebookApp.webapp_settings = {}

# Specify what command to use to invoke a web browser when opening the notebook.
# If not specified, the default browser will be determined by the `webbrowser`
# standard library module, which allows setting of the BROWSER environment
# variable to override it.
# c.NotebookApp.browser = ''

#------------------------------------------------------------------------------
# IPKernelApp configuration
#------------------------------------------------------------------------------

# IPython: an enhanced interactive Python shell.

# IPKernelApp will inherit config from: KernelApp, BaseIPythonApplication,
# Application, InteractiveShellApp

# The importstring for the DisplayHook factory
# c.IPKernelApp.displayhook_class = 'IPython.zmq.displayhook.ZMQDisplayHook'

# Set the IP or interface on which the kernel will listen.
# c.IPKernelApp.ip = '127.0.0.1'

# 
# c.IPKernelApp.parent_appname = ''

# Create a massive crash report when IPython encounters what may be an internal
# error.  The default is to append a short message to the usual traceback
# c.IPKernelApp.verbose_crash = False

# Run the module as a script.
# c.IPKernelApp.module_to_run = ''

# set the shell (ROUTER) port [default: random]
# c.IPKernelApp.shell_port = 0

# Whether to overwrite existing config files when copying
# c.IPKernelApp.overwrite = False

# Execute the given command string.
# c.IPKernelApp.code_to_run = ''

# set the stdin (DEALER) port [default: random]
# c.IPKernelApp.stdin_port = 0

# Set the log level by value or name.
# c.IPKernelApp.log_level = 30

# lines of code to run at IPython startup.
# c.IPKernelApp.exec_lines = []

# The importstring for the OutStream factory
# c.IPKernelApp.outstream_class = 'IPython.zmq.iostream.OutStream'

# Whether to create profile dir if it doesn't exist
# c.IPKernelApp.auto_create = False

# set the heartbeat port [default: random]
# c.IPKernelApp.hb_port = 0

# redirect stdout to the null device
# c.IPKernelApp.no_stdout = False

# dotted module name of an IPython extension to load.
# c.IPKernelApp.extra_extension = ''

# A file to be run
# c.IPKernelApp.file_to_run = ''

# The IPython profile to use.
# c.IPKernelApp.profile = 'default'

# Pre-load matplotlib and numpy for interactive use, selecting a particular
# matplotlib backend and loop integration.
# c.IPKernelApp.pylab = None

# kill this process if its parent dies.  On Windows, the argument specifies the
# HANDLE of the parent process, otherwise it is simply boolean.
# c.IPKernelApp.parent = 0

# JSON file in which to store connection info [default: kernel-<pid>.json]
# 
# This file will contain the IP, ports, and authentication key needed to connect
# clients to this kernel. By default, this file will be created in the security-
# dir of the current profile, but can be specified by absolute path.
# c.IPKernelApp.connection_file = ''

# If true, an 'import *' is done from numpy and pylab, when using pylab
# c.IPKernelApp.pylab_import_all = True

# The name of the IPython directory. This directory is used for logging
# configuration (through profiles), history storage, etc. The default is usually
# $HOME/.ipython. This options can also be specified through the environment
# variable IPYTHONDIR.
# c.IPKernelApp.ipython_dir = '/home/will/.config/ipython'

# ONLY USED ON WINDOWS Interrupt this process when the parent is signalled.
# c.IPKernelApp.interrupt = 0

# Whether to install the default config files into the profile dir. If a new
# profile is being created, and IPython contains config files for that profile,
# then they will be staged into the new directory.  Otherwise, default config
# files will be automatically generated.
# c.IPKernelApp.copy_config_files = False

# List of files to run at IPython startup.
# c.IPKernelApp.exec_files = []

# Enable GUI event loop integration ('qt', 'wx', 'gtk', 'glut', 'pyglet',
# 'osx').
# c.IPKernelApp.gui = None

# A list of dotted module names of IPython extensions to load.
# c.IPKernelApp.extensions = []

# redirect stderr to the null device
# c.IPKernelApp.no_stderr = False

# The Logging format template
# c.IPKernelApp.log_format = '[%(name)s] %(message)s'

# set the iopub (PUB) port [default: random]
# c.IPKernelApp.iopub_port = 0

#------------------------------------------------------------------------------
# ZMQInteractiveShell configuration
#------------------------------------------------------------------------------

# A subclass of InteractiveShell for ZMQ.

# ZMQInteractiveShell will inherit config from: InteractiveShell

# Use colors for displaying information about objects. Because this information
# is passed through a pager (like 'less'), and some pagers get confused with
# color codes, this capability can be turned off.
# c.ZMQInteractiveShell.color_info = True

# 
# c.ZMQInteractiveShell.history_length = 10000

# Don't call post-execute functions that have failed in the past.
# c.ZMQInteractiveShell.disable_failing_post_execute = False

# Show rewritten input, e.g. for autocall.
# c.ZMQInteractiveShell.show_rewritten_input = True

# Set the color scheme (NoColor, Linux, or LightBG).
# c.ZMQInteractiveShell.colors = 'Linux'

# 
# c.ZMQInteractiveShell.separate_in = '\n'

# Deprecated, use PromptManager.in2_template
# c.ZMQInteractiveShell.prompt_in2 = '   .\\D.: '

# 
# c.ZMQInteractiveShell.separate_out = ''

# Deprecated, use PromptManager.in_template
# c.ZMQInteractiveShell.prompt_in1 = 'In [\\#]: '

# Enable deep (recursive) reloading by default. IPython can use the deep_reload
# module which reloads changes in modules recursively (it replaces the reload()
# function, so you don't need to change anything to use it). deep_reload()
# forces a full reload of modules whose code may have changed, which the default
# reload() function does not.  When deep_reload is off, IPython will use the
# normal reload(), but deep_reload will still be available as dreload().
# c.ZMQInteractiveShell.deep_reload = False

# Make IPython automatically call any callable object even if you didn't type
# explicit parentheses. For example, 'str 43' becomes 'str(43)' automatically.
# The value can be '0' to disable the feature, '1' for 'smart' autocall, where
# it is not applied if there are no more arguments on the line, and '2' for
# 'full' autocall, where all callable objects are automatically called (even if
# no arguments are present).
# c.ZMQInteractiveShell.autocall = 0

# 
# c.ZMQInteractiveShell.separate_out2 = ''

# Deprecated, use PromptManager.justify
# c.ZMQInteractiveShell.prompts_pad_left = True

# 
# c.ZMQInteractiveShell.readline_parse_and_bind = ['tab: complete', '"\\C-l": clear-screen', 'set show-all-if-ambiguous on', '"\\C-o": tab-insert', '"\\C-r": reverse-search-history', '"\\C-s": forward-search-history', '"\\C-p": history-search-backward', '"\\C-n": history-search-forward', '"\\e[A": history-search-backward', '"\\e[B": history-search-forward', '"\\C-k": kill-line', '"\\C-u": unix-line-discard']

# Enable magic commands to be called without the leading %.
# c.ZMQInteractiveShell.automagic = True

# 
# c.ZMQInteractiveShell.debug = False

# 
# c.ZMQInteractiveShell.object_info_string_level = 0

# 
# c.ZMQInteractiveShell.ipython_dir = ''

# 
# c.ZMQInteractiveShell.readline_remove_delims = '-/~'

# Start logging to the default log file.
# c.ZMQInteractiveShell.logstart = False

# The name of the logfile to use.
# c.ZMQInteractiveShell.logfile = ''

# 
# c.ZMQInteractiveShell.wildcards_case_sensitive = True

# Save multi-line entries as one entry in readline history
# c.ZMQInteractiveShell.multiline_history = True

# Start logging to the given file in append mode.
# c.ZMQInteractiveShell.logappend = ''

# 
# c.ZMQInteractiveShell.xmode = 'Context'

# 
# c.ZMQInteractiveShell.quiet = False

# Deprecated, use PromptManager.out_template
# c.ZMQInteractiveShell.prompt_out = 'Out[\\#]: '

# Set the size of the output cache.  The default is 1000, you can change it
# permanently in your config file.  Setting it to 0 completely disables the
# caching system, and the minimum value accepted is 20 (if you provide a value
# less than 20, it is reset to 0 and a warning is issued).  This limit is
# defined because otherwise you'll spend more time re-flushing a too small cache
# than working
# c.ZMQInteractiveShell.cache_size = 1000

# 'all', 'last', 'last_expr' or 'none', specifying which nodes should be run
# interactively (displaying output from expressions).
# c.ZMQInteractiveShell.ast_node_interactivity = 'last_expr'

# Automatically call the pdb debugger after every exception.
# c.ZMQInteractiveShell.pdb = False

#------------------------------------------------------------------------------
# ProfileDir configuration
#------------------------------------------------------------------------------

# An object to manage the profile directory and its resources.
# 
# The profile directory is used by all IPython applications, to manage
# configuration, logging and security.
# 
# This object knows how to find, create and manage these directories. This
# should be used by any code that wants to handle profiles.

# Set the profile location directly. This overrides the logic used by the
# `profile` option.
# c.ProfileDir.location = ''

#------------------------------------------------------------------------------
# Session configuration
#------------------------------------------------------------------------------

# Object for handling serialization and sending of messages.
# 
# The Session object handles building messages and sending them with ZMQ sockets
# or ZMQStream objects.  Objects can communicate with each other over the
# network via Session objects, and only need to work with the dict-based IPython
# message spec. The Session will handle serialization/deserialization, security,
# and metadata.
# 
# Sessions support configurable serialiization via packer/unpacker traits, and
# signing with HMAC digests via the key/keyfile traits.
# 
# Parameters ----------
# 
# debug : bool
#     whether to trigger extra debugging statements
# packer/unpacker : str : 'json', 'pickle' or import_string
#     importstrings for methods to serialize message parts.  If just
#     'json' or 'pickle', predefined JSON and pickle packers will be used.
#     Otherwise, the entire importstring must be used.
# 
#     The functions must accept at least valid JSON input, and output *bytes*.
# 
#     For example, to use msgpack:
#     packer = 'msgpack.packb', unpacker='msgpack.unpackb'
# pack/unpack : callables
#     You can also set the pack/unpack callables for serialization directly.
# session : bytes
#     the ID of this Session object.  The default is to generate a new UUID.
# username : unicode
#     username added to message headers.  The default is to ask the OS.
# key : bytes
#     The key used to initialize an HMAC signature.  If unset, messages
#     will not be signed or checked.
# keyfile : filepath
#     The file containing a key.  If this is set, `key` will be initialized
#     to the contents of the file.

# Username for the Session. Default is your system username.
# c.Session.username = 'will'

# The name of the packer for serializing messages. Should be one of 'json',
# 'pickle', or an import name for a custom callable serializer.
# c.Session.packer = 'json'

# The UUID identifying this session.
# c.Session.session = ''

# execution key, for extra authentication.
# c.Session.key = b''

# Debug output in the Session
# c.Session.debug = False

# The name of the unpacker for unserializing messages. Only used with custom
# functions for `packer`.
# c.Session.unpacker = 'json'

# path to file containing execution key.
# c.Session.keyfile = ''

#------------------------------------------------------------------------------
# InlineBackend configuration
#------------------------------------------------------------------------------

# An object to store configuration of the inline backend.

# The image format for figures with the inline backend.
# c.InlineBackend.figure_format = 'png'

# Close all figures at the end of each cell.
# 
# When True, ensures that each cell starts with no active figures, but it also
# means that one must keep track of references in order to edit or redraw
# figures in subsequent cells. This mode is ideal for the notebook, where
# residual plots from other cells might be surprising.
# 
# When False, one must call figure() to create new figures. This means that
# gcf() and getfigs() can reference figures created in other cells, and the
# active figure can continue to be edited with pylab/pyplot methods that
# reference the current active figure. This mode facilitates iterative editing
# of figures, and behaves most consistently with other matplotlib backends, but
# figure barriers between cells must be explicit.
# c.InlineBackend.close_figures = True

# Subset of matplotlib rcParams that should be different for the inline backend.
# c.InlineBackend.rc = {'font.size': 10, 'savefig.dpi': 72, 'figure.figsize': (6.0, 4.0), 'figure.subplot.bottom': 0.125}

#------------------------------------------------------------------------------
# MappingKernelManager configuration
#------------------------------------------------------------------------------

# A KernelManager that handles notebok mapping and HTTP error handling

# MappingKernelManager will inherit config from: MultiKernelManager

# The max raw message size accepted from the browser over a WebSocket
# connection.
# c.MappingKernelManager.max_msg_size = 65536

# Kernel heartbeat interval in seconds.
c.MappingKernelManager.time_to_dead = 10.0

# The kernel manager class.  This is configurable to allow subclassing of the
# KernelManager for customized behavior.
# c.MappingKernelManager.kernel_manager_class = 'IPython.zmq.blockingkernelmanager.BlockingKernelManager'

# Delay (in seconds) before sending first heartbeat.
# c.MappingKernelManager.first_beat = 5.0

#------------------------------------------------------------------------------
# NotebookManager configuration
#------------------------------------------------------------------------------

# Automatically create a Python script when saving the notebook.
# 
# For easier use of import, %run and %load across notebooks, a <notebook-
# name>.py script will be created next to any <notebook-name>.ipynb on each
# save.  This can also be set with the short `--script` flag.
c.NotebookManager.save_script = True

# The directory to use for notebooks.
c.NotebookManager.notebook_dir = '/home/will/Projects/sextalk/notebooks'